docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,223 @@
+
+## Build controller
+
+[InSequence](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/sequence.go) and [InParallel](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/parallel.go) are two build controllers for deciding how to schedule the run of multiple builds together. `InSequence` runs all builds sequentially whereas `InParallel` runs them parallely with a max concurrency defined by a `concurrency` field.
+
+After introducing inter-artifact dependencies we'll need to run the builds in a topologically sorted order. One way to do this is to run any popular topologically sorting algorithm on the `Artifact` slice, treating it as a directed graph. However we can get a simpler implementation at the expense of a few additional `goroutines`.
+
+We introduce a new controller `scheduler.go` and deprecate `sequence.go` and `parallel.go`. Here we model the `Artifact` slice graph using a set of `go channels`.
+
+```go
+type artifactChanModel struct {
+	artifact              *latest.Artifact
+	artifactChan          chan interface{}
+	requiredArtifactChans []chan interface{}
+}
+
+func (a *artifactChanModel) markComplete() {
+	// closing channel notifies all listeners waiting for this build to complete
+	close(a.artifactChan)
+}
+func (a *artifactChanModel) waitForDependencies(ctx context.Context) {
+	for _, dep := range a.requiredArtifactChans {
+		// wait for dependency to complete build
+		select {
+		case <-ctx.Done():
+		case <-dep:
+		}
+	}
+}
+```
+
+Each artifact has a channel that it closes once it completes building (either success or failure) by calling `markComplete`. This notifies *all* listeners waiting for this artifact.
+
+Additionally it has a reference to the channels for each of its dependencies.
+Calling `waitForDependencies` ensures that all required artifacts' channels have already been closed and as such have finished building before the current artifact build starts.
+
+This class also provides an implementation of the interface `buildStatusRecorder` that should be safe for concurrent access.
+
+```go
+type buildStatusRecorder interface {
+  Record(imageName string, imageTag string, err error)
+  GetImageTag(imageName string) string
+}
+```
+
+Finally we have the only exported function in `scheduler.go` that orchestrates all the builds:
+
+```go
+func InOrder(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, existing []Artifact,  buildArtifact ArtifactBuilder, concurrency int) ([]Artifact, error) {
+```
+
+This function maintains an instance of `buildStatusRecorder` implementation and can pass it as an `ArtifactResolver` to the various `ArtifactBuilder`s while recording the status after each build completion.
+
+## Build logs reporting
+
+There are two quirks in the way the current `InParallel` build controller reports the build logs.
+
```

Comment:
```suggestion
Current Implementation
```

Its been long since i touched this code and first thought the below snippet it proposed code change

Reply:
added